pallets/withdraw-teleport/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+	VersionedXcm,
+};
+
+// #[cfg(test)]
+// mod mock;
+
+// #[cfg(test)]
+// mod tests;
+
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+// pub mod weights;
+// pub use weights::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_xcm::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// An error ocured during send
+		SendError,
+		/// Failed to execute
+		FailedToExecute,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Execution of an XCM message was attempted.
+		Attempted { outcome: xcm::latest::Outcome },
+		/// A XCM message was sent.
+		Sent {
+			origin: MultiLocation,
+			destination: MultiLocation,
+			message: Xcm<()>,
+			message_id: XcmHash,
+		},
+	}
+
+	/// Teleport native asset from a parachain to another.
+	/// This function is called by the parachain that wants to teleport native assets to another
+	/// parachain but needs to buy execution on the destination parachain with an asset that is not
+	/// being teleported. We call this asset the fee asset.
+	/// The parachain that wants to teleport native assets to another parachain with this method
+	/// need to fund its Sovereign Account with the fee asset on the destination parachain.
+	/// If multiple fee assets are included in the message, only the first one is used to buy
+	/// execution. Fee assets are trapped on the destination parachain.
+	/// Parameters:
+	/// - `origin`: The origin of the call.
+	/// - `dest`: The destination chain of the teleport.
+	/// - `beneficiary`: The beneficiary of the teleport from the perspective of the destination
+	///   chain.
+	/// - `native_asset_amount`: The amount of native asset to teleport.
+	/// - `fee_asset`: The fee asset to buy execution on the destination chain.
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(10_000)]
+		pub fn withdraw_and_teleport(
+			origin: OriginFor<T>,
+			dest: Box<VersionedMultiLocation>,
+			beneficiary: Box<VersionedMultiLocation>,
+			native_asset_amount: u128,
+			fee_asset: Box<VersionedMultiAssets>,
+		) -> DispatchResult {
+			Self::do_withdraw_and_teleport(
+				origin,
+				dest,
+				beneficiary,
+				native_asset_amount,
+				fee_asset,
+			)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_withdraw_and_teleport(
+		origin: OriginFor<T>,
+		dest: Box<VersionedMultiLocation>,
+		beneficiary: Box<VersionedMultiLocation>,
+		native_asset_amount: u128,
+		fee_asset: Box<VersionedMultiAssets>,
+	) -> DispatchResult {
+		//Unbox origin, destination and beneficiary.
+		let origin_location = T::ExecuteXcmOrigin::ensure_origin(origin)?;
```

Comment:
Why are you using this origin? It's not arbitrary XCM program execution and should be protected by economic means (fee payment).

Reply:
Hi! Could you please extend on this ?

Reply:
After going through the rest of the comments maybe we have different approaches on who should be able to call this ? In my mind I tried to find a way in which regular users could call it, but this is the ideal case and happy to change if doesn't make sense.

Reply:
I think I was wrong about how this is configured, I see it's the same on `do_teleport_assets`.